client.go

```diff
@@ -123,39 +146,64 @@ func init() {
 // SetUserAgent sets a custom user-agent for HTTP requests
 func (c *Client) SetUserAgent(ua string) *Client {
 	c.userAgent = ua
-	c.resty.SetHeader("User-Agent", c.userAgent)
+	c.SetHeader("User-Agent", c.userAgent)
 
 	return c
 }
 
-type RequestParams struct {
+type requestParams struct {
 	Body     any
 	Response any
 }
 
 // Generic helper to execute HTTP requests using the net/http package
 //
-// nolint:unused, funlen, gocognit
-func (c *httpClient) doRequest(ctx context.Context, method, url string, params RequestParams) error {
+// nolint:funlen, gocognit, nestif
+func (c *Client) doRequest(ctx context.Context, method, endpoint string, params requestParams, paginationMutator *func(*http.Request) error) error {
 	var (
 		req        *http.Request
 		bodyBuffer *bytes.Buffer
 		resp       *http.Response
 		err        error
 	)
 
-	for range httpDefaultRetryCount {
-		req, bodyBuffer, err = c.createRequest(ctx, method, url, params)
+	// If there is a body, ensure it's a *bytes.Reader so we can reset its position for each retry
+	var bodyReader *bytes.Reader
+	if params.Body != nil {
+		var ok bool
+		bodyReader, ok = params.Body.(*bytes.Reader)
+		if !ok {
+			return c.ErrorAndLogf("failed to read body: params.Body is not a *bytes.Reader")
+		}
+	}
```

Comment:
Just curious, if we require `params.Body` to always be `*bytes.Reader`, can we type it as `*bytes.Reader` in the `requestParams` struct?

Reply:
Yes we should be able to, but I typed it as any since that's how we had it before. Do you think it's worth updating the type?

Reply:
Since we have already made it as a private struct, I think the risk of changing it is very minimal?

CC @lgarber-akamai (for a second opinion)

https://github.com/linode/linodego/pull/576/files#diff-4b667feae66c9d46b21b9ecc19e8958cf4472d162ce0a47ac3e8386af8bbd8cfL131-R154

Reply:
Now that you mention it, I'm not sure why I changed it to be private when it was public before. I think at some point I was changing names and accidentally left it lowercase. Do you think we should keep it public like it was before or would it be better to make it private?

Reply:
I think we are fine to make a such change if we clearly document it in the release note, and I don't think this struct should necessarily be available to other packages.

Reply:
Alright sounds good! I updated the field to have a type of `*bytes.Reader` instead of `any`.